networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/LibP2PPeer.java

```diff
@@ -106,6 +144,26 @@ public void disconnectImmediately(
             error -> LOG.warn("Failed to disconnect from peer {}", getId(), error));
   }
 
+  private SafeFuture<Optional<String>> getAgentVersionFromIdentity() {
+    return getIdentify()
+        .thenApply(
+            id -> id.hasAgentVersion() ? Optional.of(id.getAgentVersion()) : Optional.empty());
+  }
+
+  private SafeFuture<IdentifyOuterClass.Identify> getIdentify() {
+    return SafeFuture.of(
+            connection
+                .muxerSession()
+                .createStream(new Identify())
+                .getController()
+                .thenCompose(controller -> SafeFuture.of(controller.id())))
```

Comment:
nit: I'd have thought you could just do
```suggestion
                .thenCompose(controller -> controller.id()))
```
and you'd still get a `SafeFuture` back from `thenCompose`